python/paddle/distributed/parallel.py

```diff
@@ -0,0 +1,297 @@
+    # is same, see [ why need set default master info before run? ]
+    args.started_port = options.get('started_port', None)
+    if args.started_port is None:
+        default_port = os.environ.get("PADDLE_MASTER_PORT", None)
+        if default_port is None:
+            raise RuntimeError(
+                "Data parallel training start failed. If you start data parallel "
+                "training by `paddle.distributed.launch` module, Please ensure "
+                "that one of the following rules is met:\n"
+                "  1. Do not set `paddle.distributed.init_parallel_env` argument "
+                "`rank` or set it to be -1;\n"
+                "  2. Set `paddle.distributed.init_parallel_env` start port for "
+                "parallel training by `started_port=**`, e.g. started_port=6170."
+            )
+        args.started_port = int(default_port)
+
+    args.use_paddlecloud = options.get('use_paddlecloud', False)
+    args.print_config = options.get('print_config', True)
+
+    # set default `selected_gpus`
+    # TODO(chenweihang): if users gived number of `selected_gpus`
+    # is not equal to the spawn's nprocs, it will cause error, 
+    # and because we remove the `proc num` argument of 
+    # `init_parallel_env`, when above error occured, we do not 
+    # have a good way to check, so users are not recommended to 
+    # use this parameter, it is best to delete
+    args.selected_gpus = options.get('selected_gpus', None)
+    if args.selected_gpus is None:
+        args.selected_gpus = os.environ.get("PADDLE_CUDA_VISIBLE_DEVICES", None)
+        if args.selected_gpus is None:
+            raise ValueError(
+                "Data parallel training start failed. If you start data parallel "
+                "training by `paddle.distributed.launch` module, Please ensure "
+                "that one of the following rules is met:\n"
+                "  1. Do not set `paddle.distributed.init_parallel_env` argument "
+                "`rank` or set it to be -1;\n"
+                "  2. Set `paddle.distributed.init_parallel_env` selected gpus of "
+                "parallel training by `selected_gpus=**`, e.g. selected_gpus='0,1,2,3'."
+            )
+
+    # reuse code of launch.py
+    cluster, pod = get_cluster_and_pod(args)
+
+    # remove useless env vars
+    os.environ.pop("http_proxy", None)
+    os.environ.pop("https_proxy", None)
+
+    # update env vars
+    trainer = pod.get_trainer(rank)
+    if trainer is None:
+        raise RuntimeError(
+            "The expected trainer is not exists, its trainer rank is %d." %
+            rank)
+    proc_env = {
+        "FLAGS_selected_gpus": "%s" % ",".join([str(g) for g in trainer.gpus]),
+        "PADDLE_TRAINER_ID": "%d" % trainer.rank,
+        "PADDLE_CURRENT_ENDPOINT": "%s" % trainer.endpoint,
+        "PADDLE_TRAINERS_NUM": "%d" % cluster.trainers_nranks(),
+        "PADDLE_TRAINER_ENDPOINTS": ",".join(cluster.trainers_endpoints())
+    }
```

Comment:
As so much code is copied from launch.py, should there be a utility file reserving the common code which is shard by both parallel.py and launch.py?

Reply:
thanks, I extract the `get_cluster_and_pod` method to avoid most of the code rewriting, but as you say the `proc_env` update code is still duplicate, I will also extract this function later

Reply:
done, thx